Cherry-picking @aaime works on SLD

rldhont opened the pull request:
Do not generate a ogc:Filter where a ogc:Expression is expected. #3366
Emit valid SLD when mm are used as symbol unit #3414

aaime committed:
Do not generate a ogc:Filter where a ogc:Expression is expected. Fixes #10076 and #11202

aaime committed:
Emit valid SLD when on screen mm are used

SLD does not have a notion of on screen mm, rescale them to pixels to get an equivalent, more valid and more widely usable, output

aaime committed:
Export map level scale based dependencies in most vector symbology

aaime commented:
Hey, thanks a ton for backporting. As-is this one will introduce a segfault in QGis, I did not notice until I worked on the third, non backportable as-is pull request I made (the one about scale dependencies).
The fix is small, fully contained in qgissymbollayerutils.*, it's the changes in how the rotation and alpha are encoded, using a new createExpressionElement dedicated to expressions.

Initially I mistakenly though createFunctionElement was used to encode only functions/expressions, but it turns out it's also used to encode filters, so I un-did the changes in there and created a new dedicated method to encode expressions only.
See in particular:
https://github.com/qgis/QGIS/pull/3436/files#diff-1f3d1469bf313b6e4de42a116952f95fR227
https://github.com/qgis/QGIS/pull/3436/files#diff-90992a7558fbe4a11528130ebbf70987R2156
(besides the methods handling scale dependencies)

elpaso commented:
@aaime can you fix this PR and make a new one that will not segfault QGIS?

aaime commented:
@elpaso I'll try during one of the incoming weekends, but no promises :-)

rldhont committed:
Fix cherry-pickings

Do not generate a ogc:Filter where a ogc:Expression is expected. #3366
Emit valid SLD when mm are used as symbol unit #3414
Export map level scale based dependencies in most vector symbology #3436

rldhont commented:
I have cherry-picking: Export map level scale based dependencies in most vector symbology #3436, and fixing them.

aaime commented:
Hi @rldhont sorry for not being active on this one, I have spent time getting the parametric SVGs support in GeoTools before the feature freeze the incoming weekend. I've been told that the scale dependency one is not backportable as implemented because I changed an existing virtual method, but believe that adding a parallel one would be backportable instead. @m-kuhn and @nyalldawson can surely provide better explanation than me.

rldhont commented:
All tests passed!
@elpaso @m-kuhn and @aaime is it ready to merge ?

m-kuhn commented:
@rldhont this still is an API change, overrides will no longer trigger:

https://github.com/qgis/QGIS/pull/3448/files#diff-7104e041f14ce3bd399e6be00f3a1a48R102

Instead making a new method with the additional parameter which internally calls the old method if not overwritten by subclasses should be ok.

rldhont commented:
@m-khun something like that:
```
    virtual void toSld( QDomDocument& doc, QDomElement &element, QgsStringMap props = QgsStringMap() ) const override;
    virtual void toSld( QDomDocument& doc, QDomElement &element, QgsStringMap props = QgsStringMap() ) { Q_UNUSED( props ); toSld(doc, element); }
```

m-kuhn commented:
@rldohnt ;)

    Q_DECL_DEPRECATED virtual void toSld( QDomDocument& doc, QDomElement &element ) const;

    virtual void toSld( QDomDocument& doc, QDomElement &element, QgsStringMap props )
    {
      Q_UNUSED( props );
      toSld(doc, element);
    }

elpaso commented:
@rldhont I'll check this ASAP

rldhont committed:
Keep API compatibility and update SIP

elpaso commented:
@rldhont your last commit LGTM but to pass the tests you should probably add something like
https://github.com/qgis/QGIS/blob/master/src/core/qgspallabeling.h#L140
 to the deprecated methods.

rldhont commented:
Thanks @elpaso, but I use which number 2.17 or 3.0 ?
And If I'd like to backport it to 2.14 and 2.16 ?

m-kuhn commented on python/core/qgsvectorlayer.sip:
```diff
@@ -564,7 +564,7 @@ class QgsVectorLayer : QgsMapLayer
     bool writeStyle( QDomNode& node, QDomDocument& doc, QString& errorMessage ) const;
 
-    bool writeSld( QDomNode& node, QDomDocument& doc, QString& errorMessage ) const;
+    bool writeSld( QDomNode& node, QDomDocument& doc, QString& errorMessage, QgsStringMap& props ) const;
```
This is an API break, use this instead

    bool writeSld( QDomNode& node, QDomDocument& doc, QString& errorMessage, const QgsStringMap& props = QgsStringMap() ) const;

m-kuhn commented on python/core/symbology-ng/qgsgraduatedsymbolrendererv2.sip:
```diff
@@ -108,7 +108,7 @@ class QgsGraduatedSymbolRendererV2 : QgsFeatureRendererV2
     virtual QgsGraduatedSymbolRendererV2* clone() const /Factory/;
 
-    virtual void toSld( QDomDocument& doc, QDomElement &element ) const;
+    virtual void toSld( QDomDocument& doc, QDomElement &element, QgsStringMap props ) const;
```
This is an API break, any virtual method signature must not be changed.

elpaso replied:
What is the best solution here? Add a default value to props? 
```
virtual void toSld( QDomDocument& doc, QDomElement &element, QgsStringMap props = QStringMap( ) ) const;
 virtual void toSld( QDomDocument& doc, QDomElement &element ) const;
```

Or we need to leave the old signature in the bindings? (I hoped python could use the new signature since the additional parameter is optional)

rldhont replied:
Do I have to add both signatures ?

m-kuhn reviewed (changes requested):
There are still API breaks in there. Basically:

Any virtual method signature must be left untouched.
Any non-virtual method can be extended with additional parameters if these additional parameters have a default value.

elpaso commented:
@rldhont that's a good question, I think the deprecation applies to the version we are working on now, so it would be different in the different branches, @m-kuhn what do you think?

m-kuhn commented:
I am not sure this is suitable as a backport, I find it hard to treat this as a bugfix rather than a new feature, no?

rldhont commented:
@m-kuhn it's a hard bugfix, but it's a bugfix in the interoperability point of view. I don't want to force the backport to 2.14 but it's our LTR and I think it's a good enhancement for QGIS Server.

m-kuhn commented:
How is QGIS Server affected by this?

elpaso commented:
@m-kuhn IMHO this is a bugfix because the SLD exported from QGIS is not validating as XML and leads to severe compatibility problems with other software.

rldhont commented:
QGIS Sever uses the same method as QGIS to export SLD in the GetStyles request. This request is made to share styles in an OGC manner.

m-kuhn commented on src/core/symbology-ng/qgsrendererv2.h:
```diff
@@ -238,7 +238,7 @@ class CORE_EXPORT QgsFeatureRendererV2
     //! create the SLD UserStyle element following the SLD v1.1 specs with the given name
     //! @note added in 2.8
-    virtual QDomElement writeSld( QDomDocument& doc, const QString& styleName ) const;
+    virtual QDomElement writeSld( QDomDocument& doc, const QString& styleName, QgsStringMap props = QgsStringMap() ) const;
```
The main issue is here, you need to add a new method instead of adding a new parameter (it's virtual)

m-kuhn commented on src/core/symbology-ng/qgsrendererv2.h:
```diff
@@ -253,8 +253,15 @@ class CORE_EXPORT QgsFeatureRendererV2
+    virtual void toSld( QDomDocument& doc, QDomElement &element, QgsStringMap props = QgsStringMap() ) const
+    {
+      element.appendChild( doc.createComment( QString( "FeatureRendererV2 %1 not implemented yet" ).arg( type() ) ) );
+      ( void ) props; // warning avoidance
```
Use `Q_UNUSED`

m-kuhn commented on src/core/symbology-ng/qgsrendererv2.h:
```diff
@@ -253,8 +253,15 @@ class CORE_EXPORT QgsFeatureRendererV2
+    {
+      toSld( doc, element, QgsStringMap() );
+    }
+    virtual void toSld( QDomDocument& doc, QDomElement &element, QgsStringMap props = QgsStringMap() ) const
```
I wonder, isn't this ambiguous? What will be called with

    toSld( doc, element );

The new method with default parameter or the old method?

rldhont replied:
Has virtual method, what do you recommend ?

m-kuhn replied:
I was just wondering how this is handled. I would probably have introduced the new method without default value for the new parameter to avoid confusion.

m-kuhn commented:
If it's backported I'd say we don't deprecate methods there.

rldhont commented:
@elpaso @m-kuhn I have updated API break:
* remove depreacated for backporting
* duplicature virtual signature
* set props as const QgsMapString&
Thanks for your reviews

rldhont commented:
@m-khun and @elpaso the build for qt4 is OK
Before merging, which note added can I write ? If I'd like to backport ?

m-kuhn commented:
@rdlhont you need to add a couple of `override`s to the overridden members to get the qt5 tests to pass.
Just add the note for the oldest version to which it is backported.

rldhont commented:
All checks have passed :) Can I merge and backport ? @elpaso @m-kuhn

m-kuhn commented:
Looks good to me.
Thanks @rldhont !!

elpaso commented:
I trust @m-kuhn :) Thanks @rldhont  for the good job!

m-kuhn commented:
I just checked the api compat. Functionalitywise i rely on your testing ;)

On September 20, 2016 4:14:31 PM GMT+02:00, Alessandro Pasotti <notifications@github.com> wrote:
>I trust @m-kuhn :) Thanks @rldhont  for the good job!
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/qgis/QGIS/pull/3448#issuecomment-248313674

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

rldhont commented:
Thanks, I merge.
